### ISSUE TITLE
stability: consistency checker's diff is hard to parse

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#9554]
The current format is as follows.

```
k:v = (/Local/Range/"\xbe\x89\x12T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90\x00\x01\x89\x89"/Transaction/addrKey:/id:"47328368-d6f1-4b07-bde2-03117cacea26" (/Local/Range/"\xbe\x89\x12T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90\x00\x01\x89\x89"/Transaction/addrKey:/id:"47328368-d6f1-4b07-bde2-03117cacea26"), 0.000000000,0, 12040800100018002000280032b2012103fde7030a460a1047328368d6f14b07bde203117cacea2610001a17be891254efc22240654793a18f9a7e54d297900001898920002a0c08d2c58aa7e1cbd9bb14100030d590d80238064000121b73716c2f6578656375746f722e676f3a3439392073716c2074786e2000320c08d2c58aa7e1cbd9bb1410003a0c08c188bba7e1cbd9bb14100042100806120c08c188bba7e1cbd9bb14100042100809120c08d2c58aa7e1cbd9bb141000480160006800) not present on lease holder
k:v = (/Local/Range/"\xbe\x89\x12\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe.\x00\x01\x89\x89"/Transaction/addrKey:/id:"ada5cc60-257c-43cc-ae65-f814b039bc94" (/Local/Range/"\xbe\x89\x12\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe.\x00\x01\x89\x89"/Transaction/addrKey:/id:"ada5cc60-257c-43cc-ae65-f814b039bc94"), 0.000000000,0, 12040800100018002000280032b201485d2fb8030a460a10ada5cc60257c43ccae65f814b039bc9410001a17be8912860a3441f4ca45929e67429e33a2fe2e0001898920002a0c08faa68ea8e1cbd9bb14100030f2bba40138064000121b73716c2f6578656375746f722e676f3a3439392073716c2074786e2000320c08faa68ea8e1cbd9bb1410003a0c0883acb1a8e1cbd9bb14100042100805120c08faa68ea8e1cbd9bb14100042100806120c0883acb1a8e1cbd9bb141000480160006800) not present on lease holder
k:v = (/Local/Range/"\xbe\x89\x12\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|\x00\x01\x88"/Transaction/addrKey:/id:"d4d2bdd5-8467-45b4-bbbc-8c20b523e3a6" (/Local/Range/"\xbe\x89\x12\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|\x00\x01\x88"/Transaction/addrKey:/id:"d4d2bdd5-8467-45b4-bbbc-8c20b523e3a6"), 0.000000000,0, 12040800100018002000280032b1018168ce1a030a450a10d4d2bdd5846745b4bbbc8c20b523e3a610001a16be8912882f29bda1144195a42fcccb6b40297c00018820002a0c08ded195a7e1cbd9bb14100030beb1e60238034000121b73716c2f6578656375746f722e676f3a3439392073716c2074786e2000320c08ded195a7e1cbd9bb1410003a0c08ccf7afa7e1cbd9bb14100042100804120c08ded195a7e1cbd9bb14100042100806120c08ccf7afa7e1cbd9bb141000480160006800) not present on lease holder
```

It has the following shortcomings:
- the interesting bit of information (whether it's on the lease holder or the follower) is at the end of the overly long line
- the key is printed twice. The intention is that the second time around it's printex as hex, but that isn't working.
- the timestamps are tucked away awkwardly inbetween.

I would suggest the following format:

```
X <ts> <pretty_key>
  value:<pretty_value>
  raw_key:<hex_key> raw_value:<hex_value>
```

where `X` is either `F` (this kv is on the follower, but not on the lease holder) or `L` (the opposite).
It's fairly straightforward to grep through (`grep -E '[^ ]'` keeps only the first line, `grep -F ' value:'` gives the second line, etc) and there it's less overwhelming to look at.

See #9531.

**Comment by tschottdorf** [comment]
After manually reformatting the output above for #9531, some additions:

- the timestamp field should be aligned (filled up) so that the `<pretty_key>` column is aligned
- `L` and `F` should be replaced by `-` and `+` because that gives us free diffs. Perhaps an explanatory line at the beginning of the output to clarify what each means (in the diff below, `-` means "this was on the proposer, but not on the follower", which I believe makes sense).

This is what the result would look like:

```diff
- 0.000000000,0           /Local/RangeID/4406/r/RangeStats
+ 0.000000000,0           /Local/RangeID/4406/r/RangeStats
+ 0.000000000,0           /Local/Range/"\xbe\x89\x12T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90\x00\x01\x89\x89"/Transaction/addrKey:/id:"47328368-d6f1-4b07-bde2-03117cacea26"
+ 0.000000000,0           /Local/Range/"\xbe\x89\x12\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe.\x00\x01\x89\x89"/Transaction/addrKey:/id:"ada5cc60-257c-43cc-ae65-f814b039bc94"
+ 0.000000000,0           /Local/Range/"\xbe\x89\x12\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|\x00\x01\x88"/Transaction/addrKey:/id:"d4d2bdd5-8467-45b4-bbbc-8c20b523e3a6"
+ 0.000000000,0           /Local/Range/"\xbe\x89\x12\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5\x00\x01\x89\x89"/Transaction/addrKey:/id:"113cf6cd-383c-4f42-aba9-c0d7bf818265"
+ 0.000000000,0           /Local/Range/"\xbe\x89\x12\x91\xc8\x16\xa2\x87\xa1G\xfa\x83\x1a\xf21?\r.\x00\x01\x88"/Transaction/addrKey:/id:"401e3c50-502e-4404-a092-91904bdd0c74"
- 1474692245.960286137,1  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 0.000000000,0           /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 1474759957.250482898,0  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 1474692245.960286137,1  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
- 1474668615.847633162,0  /Table/54/1/"\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe."/1/1
+ 0.000000000,0           /Table/54/1/"\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe."/1/1
+ 1474759957.252641658,0  /Table/54/1/"\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe."/1/1
+ 1474668615.847633162,0  /Table/54/1/"\x86\n4A\xf4\xcaE\x92\x9egB\x9e3\xa2\xfe."/1/1
+ 0.000000000,0           /Table/54/1/"\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|"/0
+ 1474759957.250664670,0  /Table/54/1/"\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|"/0
+ 0.000000000,0           /Table/54/1/"\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|"/1/1
+ 1474759957.250664670,0  /Table/54/1/"\x88/)\xbd\xa1\x14A\x95\xa4/\xcc\xcbk@)|"/1/1
- 1474766141.643815777,0  /Table/54/1/"\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5"/1/1
- 1474651012.586033417,1  /Table/54/1/"\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5"/1/1
+ 0.000000000,0           /Table/54/1/"\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5"/1/1
+ 1474759957.246878299,0  /Table/54/1/"\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5"/1/1
+ 1474651012.586033417,1  /Table/54/1/"\x8b\xce{m\xa5.G\xf8\x83\xd6J`\x0e\x03\x92\xe5"/1/1
+ 0.000000000,0           /Table/54/1/"\x91\xc8\x16\xa2\x87\xa1G\xfa\x83\x1a\xf21?\r."/0
+ 1474759957.253245522,0  /Table/54/1/"\x91\xc8\x16\xa2\x87\xa1G\xfa\x83\x1a\xf21?\r."/0
+ 0.000000000,0           /Table/54/1/"\x91\xc8\x16\xa2\x87\xa1G\xfa\x83\x1a\xf21?\r."/1/1
+ 1474759957.253245522,0  /Table/54/1/"\x91\xc8\x16\xa2\x87\xa1G\xfa\x83\x1a\xf21?\r."/1/1
```

**Comment by tschottdorf** [comment]
Other improvements here were also pointed out in #9531 but are pretty nicely visible above when looking at the `1474692245.960286137,1 ` line. It is present on both, but due to other values mismatching inbetween, it is erroneously displayed twice. This shouldn't happen. In other words, the block

```diff
- 1474692245.960286137,1  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 0.000000000,0           /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 1474759957.250482898,0  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 1474692245.960286137,1  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
```

should simply read

```diff
+ 0.000000000,0           /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
+ 1474759957.250482898,0  /Table/54/1/"T\xef\xc2\"@eG\x93\xa1\x8f\x9a~T?\x90"/1/1
```

**Comment by tschottdorf** [comment]
One more remark is that it also makes sense to include the pretty timestamp, but out of band:

```diff
+1474766141.643815777,0 /Local/RangeID/4406/r/RangeStats
+  ts:2016-09-25 01:15:41.643815777 +0000 UTC
+  value:<pretty_value>
+  raw_key:<hex_key> raw_value:<hex_value>
```

The pretty time above is `fmt.Println(time.Unix(0, ts.WallTime))` and should be `<zero>` for the zero timestamp (or the line simply omitted). It's worth considering a terser format for nontrivial timestamps, too.

**Comment by tschottdorf** [comment]
The item left here is to pretty-print the value. I didn't do that in #9557 since it requires lifting some code from the `cli/debug.go` realm and it wasn't trivial. Help here would be appreciated.

**Comment by tschottdorf** [comment]
Good enough for now.

**Comment by tamird** [comment]
I'm going to reopen this. It's a decent new contributor issue.

**Comment by tschottdorf** [comment]
Maybe. The way I thought this would be addressed is by making pretty-printing value-aware, i.e. change the API of the pretty printer to something like

```go
func PrettyPrint(key roachpb.Key, valueBytes []byte) (string, proto.Message) {
  // Do only the current thing if `len(valueBytes)==0`, otherwise unmarshal valueBytes into
  // a proto of the type we know the key is used for.
}
```

which makes this a larger refactor (and in particular removes all the tryX methods in cli/debug.go as well as some code that pretty-prints in SQL).
